db.go

```diff
@@ -309,6 +297,58 @@ func Open(path string, mode os.FileMode, options *Options) (*DB, error) {
 	return db, nil
 }
 
+// getPageSize reads the pageSize from the meta pages. It tries
+// to read the first meta page firstly. If the first page is invalid,
+// then it tries to read the second page using the default page size.
+func (db *DB) getPageSize() (int, error) {
+	var (
+		meta0CanRead, meta1CanRead bool
+	)
+
+	// Read the first meta page to determine the page size.
+	if pgSize, canRead, err := db.getPageSizeFromMetadata(0); err != nil {
+		// We cannot read the page size from page 0, but can read page 0.
+		meta0CanRead = canRead
+	} else {
+		return pgSize, nil
+	}
+
+	// Read the second meta page to determine the page size.
+	if pgSize, canRead, err := db.getPageSizeFromMetadata(1); err != nil {
+		// We cannot read the page size from page 1, but can read page 1.
+		meta1CanRead = canRead
+	} else {
+		return pgSize, nil
+	}
+
+	// If we can't read the page size from both pages, but can read
+	// either page, then we assume it's the same as the OS or the one
+	// given, since that's how the page size was chosen in the first place.
+	//
+	// If both pages are invalid, and (this OS uses a different page size
+	// from what the database was created with or the given page size is
+	// different from what the database was created with), then we are out
+	// of luck and cannot access the database.
+	if meta0CanRead || meta1CanRead {
+		return db.pageSize, nil
+	}
+
+	return 0, ErrInvalid
+}
+
+// getPageSizeFromMetadata reads the pageSize from the given meta page id
+func (db *DB) getPageSizeFromMetadata(id pgid) (int, bool, error) {
+	var buf [0x1000]byte
+	var metaCanRead bool
+	if bw, err := db.file.ReadAt(buf[:], int64(id*pgid(db.pageSize))); err == nil && bw == len(buf) {
```

Comment:
Isn't this a chicken & egg problem ? 
To read the second page... we need to know the size of the pages.

Reply:
Alternatives:
  a) search for 'const magic uint32 = 0xED0CDAED' in the prefix (10MB?) of the file 
and consider position of the second occurance as candidate for the page size. 
Note: It might be an attack vector if someone will start writing to the DB 0xED0CDAED to trick the page-sizes. 
 
  b) Try different page sizes exponentially e.g. from 1KB to 8MB ?
  
 [I mention ~MB pages,  assuming eventually bbolt might be usable with https://wiki.debian.org/Hugepages]

Reply:
Fixed.